### PR TITLE
feat(google_cloud): instrumentation for google bigquery

### PR DESCRIPTION
**Pull request by liadgo**

### CODE REVIEW

File: src/events/google_cloud.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+const tryRequire = require('../try_require.js');
+const utils = require('../utils.js');
+const tracer = require('../tracer.js');
+const serverlessEvent = require('../proto/event_pb.js');
+const eventInterface = require('../event.js');
+const errorCode = require('../proto/error_code_pb.js');
+
+const common = tryRequire('@google-cloud/common/');
+
+const uriSplitString = 'googleapis.com/';
```
Comment (galbash) [inline src/events/google_cloud.js]: Const should be all caps i think (look at other modules

File: src/events/google_cloud.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+ * @returns {Function} The wrapped function
+ */
+function googleWrapper(wrappedFunction) {
+    return function internalOWWrapper(reqOpts, config, callback) {
+        const uri = reqOpts.uri.split(uriSplitString)[1];
+        const splitUri = uri.split('/');
+        const service = splitUri[0];
+        const projectId = splitUri[3];
+        const path = uri.split(`${projectId}/`)[1];
+        const operation = path.split('/')[0];
```
Comment (galbash) [inline src/events/google_cloud.js]: Al these assignments seems very opportunistic (past  7+- lines), nothing can go wrong here? Do we want some default values?

File: src/events/google_cloud.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        const resource = new serverlessEvent.Resource([
+            projectId,
```
Comment (galbash) [inline src/events/google_cloud.js:31]: Is this the resource name? We probably want a better name

File: src/events/google_cloud.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+            service,
+            operation,
+        ]);
+        const startTime = Date.now();
+        const invokeEvent = new serverlessEvent.Event([
+            `${service}-${uuid4()}`,
```
Comment (galbash) [inline src/events/google_cloud.js]: Is there no google messagi id or something? Per service, might be useful for later strong id connections

File: src/events/google_cloud.js [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+            service,
+            0,
+            errorCode.ErrorCode.OK,
+        ]);
+
+        invokeEvent.setResource(resource);
+        if (reqOpts.json !== undefined) {
+            eventInterface.addToMetadata(
+                invokeEvent,
+                reqOpts.json
```
Comment (galbash) [inline src/events/google_cloud.js]: Do it as the next parameter maybe? Not to add on metadata only

File: src/events/google_cloud.js [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+            );
+        }
+
+        let patchedCallback;
+        const responsePromise = new Promise((resolve) => {
+            patchedCallback = (err, body, response) => {
+                invokeEvent.setDuration(utils.createDurationTimestamp(startTime));
+                eventInterface.addToMetadata(invokeEvent, response.body);
+                callback(err, body, response);
+                resolve();
```
Comment (galbash) [inline src/events/google_cloud.js:68]: Resolve before calling the callback (or we might never resolve on exceptions)

**Review by galbash — CHANGES_REQUESTED** [review]
Looks really good, few small comments

File: src/events/google_cloud.js (hunk 1 above)
Comment (liadgo) [reply src/events/google_cloud.js]: K

File: src/events/google_cloud.js (hunk 6 above)
Comment (liadgo) [reply src/events/google_cloud.js:68]: OK

File: src/events/google_cloud.js (hunk 2 above)
Comment (liadgo) [reply src/events/google_cloud.js]: :)

File: src/events/google_cloud.js (hunk 3 above)
Comment (liadgo) [reply src/events/google_cloud.js:31]: Talked about it

**Review by galbash — DISMISSED** [review]

**Review by razlo — APPROVED** [review]

**Comment by ranrib** [comment]
fixes https://github.com/epsagon/epsagon-node/issues/178

**Pull request merged**